src/NUnitFramework/framework/Constraints/CollectionTally.cs

```diff
@@ -69,31 +69,14 @@ public CollectionTallyResult Result
         /// <param name="c">The expected collection to compare against.</param>
         public CollectionTally(NUnitEqualityComparer comparer, IEnumerable c)
         {
-            this.comparer = comparer;
-
-            foreach (object o in c)
-                _missingItems.Add(o);
-        }
-
-        private bool ItemsEqual(object expected, object actual)
-        {
-            Tolerance tolerance = Tolerance.Default;
-            return comparer.AreEqual(expected, actual, ref tolerance);
+            _missingItems = new MissingItemCollection(comparer, c);
         }
 
         /// <summary>Try to remove an object from the tally.</summary>
         /// <param name="o">The object to remove.</param>
         public void TryRemove(object o)
         {
-            for (int index = _missingItems.Count - 1; index >= 0; index--)
-            {
-                if (ItemsEqual(_missingItems[index], o))
-                {
-                    _missingItems.RemoveAt(index);
-                    return;
-                }
-            }
-
+            if (_missingItems.Remove(o)) return;
```

Comment:
If we are doing a non-transitive equality comparison, such as when the tolerance is non-zero, we can't use _missingItems because _missingItems is based on a hash table (Dictionary).

Reply:
In fact, the `TryRemove(object)` paradigm is fundamentally broken if we are using a non-transitive comparison, so you're taking us from one kind of broken code to another kind of broken code. @nunit/framework-team Is this okay? Maybe we can deal with non-transitivity properly in a separate effort since it looks like it has always been broken.

Reply:
The current collection equality compare operations do not support non-transitive comparisons. If transitive collection comparisons were required, interval trees might be used to improve performance.

Reply:
@rprouse, @CharliePoole: I feel like we need more input than just me here. Is it okay for EquivalentTo with a tolerance to go from one kind of broken to the same kind of broken but with different side effects?

Reply:
@jnm2 Since you were going in a different direction, I had bowed out of this one. I can take another look but it will be easier for me (maybe others too) if you summarize how you see the existing collection tally as "broken" and what new firm of "broken" you see as coming in to replace it. Could be the limitations of following this on my cell, but I can't see what you are getting at.

If my memory is correct, we don't support tolerance in equivalence operations right now. Are you trying to add that support while working on performance improvement? If so, I'd say let's pick one goal at a time to work on even if we have to visit the same code twice.

Reply:
@CharliePoole I wasn't aware of the tolerance issue until https://github.com/nunit/nunit/issues/2799#issuecomment-379943712, and we discussed what that would look like in the comments immediately following that.

I don't believe I said that CollectionTally is broken; "EquivalentTo with a tolerance" is what I said was broken. Our equivalence implementation takes optimization shortcuts which are invalid when a tolerance is specified. We don't report an error if a user tries to do it. We silently return the wrong answer instead. Here is an example of something that fails which should succeed:

```cs
Assert.That(
    new[] { 2f, 1f },
    Is.EquivalentTo(new[] { 3f, 2f })
        .Using<float>((a, b) => Math.Abs(a - b) <= 1) );
```

Are you open to the idea making it an error to use collection equivalence with a non-transitive comparison? I like that as I said above because it allows us to implement that separately in the future, if at all.
If you think there might be folks doing `Is.EquivalentTo(...).Using(...)` today though, is it okay to take that semi-functional behavior away from them? Should we provide a thorough fix?

Reply:
@jnm2 You mentioned me and asked for input. I haven't been following this, so I'm just trying to clarify what you see as broken.  CollectionTally was mentioned previously so I guessed that was it.

We decided not to support tolerance in assertions that compare collections as sets because it's a pretty difficult problem of optimization with the potential of not converging to a solution. It's possible to reverse the decision but I wouldn't call that broken.

Another past decision was to allow folks to specify a tolerance in cases where we will not actually use it. It's similar to how you can specify case insensitive and then do a numeric test. That decision could also be changed or we could simply stop allowing it on certain constraints.q

I thought that this issue was aimed at improving efficiency without changing behavior. I think that's a better approach so since you asked I'm telling you, even though it's you who will have to move forward with the consequences.

Totally on the meta side but related... I was just in a very interesting session with Kent Beck around the question of what happens as diffs become smaller and smaller. At some point. He thin

Reply:
@CharliePoole This is why EquivalentTo, in combination with some of its `Using` methods, is broken:

- `Is.EquivalentTo(...).Using<double>((a, b) => Math.Abs(a - b) <= 1)` doesn't cause an error.
  I'd expect an InvalidOperationException if we decided not to support collection asserts with tolerances.

- `Is.EquivalentTo(...).Using<double>((a, b) => Math.Abs(a - b) <= 1)` doesn't ignore the custom comparer, either. It returns answers that are different from `Is.EquivalentTo(...)`, answers which are not only different from `Is.EquivalentTo(...)` but also which are incorrect for `Is.EquivalentTo(...).Using(...)`.

It's not documented to the user as such, but this constitutes undefined behavior. Our optimizations make it very order-sensitive. It will usually appear to the user that `Is.EquivalentTo(...).Using(...)` is working as intended while the user is setting up the test.

Wouldn't the right thing be to throw InvalidOperationException?
If we don't, this PR will have the side effect of changing the behavior of `Is.EquivalentTo(...).Using(...)` from something which appears to work as intended but doesn't, to something which still appears to work but fails to work as intended in a different way.

Reply:
Yes I said that allowing using is a potential bug. It's a bug because the user can simulate use of a tolerance as you do. I suspect that it will take a bit of discussion to select the best way to fix it.

As a bug, I feel it needs an issue, decision and fix. If I were doing it, I'd fix it first, followed by the performance fix. It doesn't seem like it should take long once you decide. It's a choice between disallowing it at compile time or throwing an exception later.

If that's your sole concern, I'm with you. It sounded like you were faulting us for not implementing equivalence with tolerance, which I think was correct.

Reply:
@CharliePoole I'm sorry I came across as faulting you for anything, particularly a decision I agree with. I referred to tolerance when I meant the more inclusive term 'non-transitivity,' so I probably confused both of us!

> I thought that this issue was aimed at improving efficiency without changing behavior. I think that's a better approach.

You mentioned that you thought this issue was aimed at improving efficiency without changing behavior. (Specifically, the behavior of `Is.EquivalentTo(...).Using(...)` will be scrambled around with a hash table.)
My point is that this issue *can't* improve efficiency without changing behavior, unless this PR waits for a separate PR to throw InvalidOperationException. This is my sole concern.

The meta comment is interesting to me. If you didn't use branches, would you still keep the master branch continuously deliverable?

Reply:
@CharliePoole I'm sorry I came across as faulting you for anything, and ironically a decision I agree with. I referred to tolerance when I meant the more inclusive term 'non-transitivity,' so I probably confused both of us!

You mentioned that you thought this issue was aimed at improving efficiency without changing behavior. 
This is my sole concern: this issue *can't* improve efficiency without changing behavior, unless this PR waits for a separate PR to throw InvalidOperationException. (Specifically, the behavior of `Is.EquivalentTo(...).Using(...)` will be scrambled around with a hash table.)

The meta comment is interesting to me. If you didn't use branches, would you still keep the master branch continuously deliverable?

Reply:
Here's the discussion issue: https://github.com/nunit/nunit/issues/2870

Reply:
@jnm2 More of a problem of clarity for me and it's clear now. nowv

Reply:
Regarding the meta comment, I don't think any of us fully understood but I think so. If you had been there and asked he'd have likely said "Go figure that out" 'cause that's who he is.

His idea includes some special tools that don't exist yet as well as putting an AST in the repo rather than text, so it's kind of futuristic.

My idea is a simpler subset... what if diffs were so small that you could automatically commit them once the tests passed?

Reply:
If you automatically commit when tests pass, are there still reviews and discussions?

Reply:
I might decide NUnit should have a brand new feature or remove an API. I'd still seek out the team's opinions and then want to link the implementation back to the discussion, right?

Reply:
I'd sure think so.

Reply:
If reviews were eliminated, how would the team handle mistakes if my commit didn't quite match the discussion? Master would then not be deliverable.

Reply:
Of course any change can be reversed.

However. You aren't getting in the "what if" spirit. The question was how small would a change have to be to eliminate all review. Or what would have to be true about the change?

What about a small refactoring? A large one? A single new failing test plus code to make it pass? A bunch of new tests that improve coverage and pass without changes?

Finally, if we could do a lot of work more simply and quickly, how would it affect the way we release?

What if is powerful. Somebody was the first to sdk What if each dev could have his own private repo for example.

Reply:
I'll keep thinking about it.